Add elasticsearch alerting engine

beardface opened the pull request:
#5893 

This branch introduces a, hopefully, lighter-weight variant of an elasticsearch alerting engine for Grafana.

The other approaches I've seen seem very heavy on implementation; and internal ES logic.  This approach attempts to take what I consider to be a simpler/cleaner path.

It is working well for us; so I figured I'd share in the event its useful enough to merge; or others would like to leverage the implementation.

beardface committed:
initial es alerting

beardface committed:
get es working

beardface committed:
improve test coverage and formatting

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=7551) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=7551) before we can accept your contribution.<br/><hr/>**Justin Hawkins** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

beardface committed:
add vendored files

beardface committed:
remaining vendor files

beardface committed:
remaining vendor files

beardface committed:
Merge branch 'esAlerting' of https://github.com/RiotGames/grafana into esAlerting

torkelo commented:
looks like a good start, do we really need the elastic go client lib? 

Also the response handling looks a bit simplistic, does it support multiple levels of bucket aggregations? For example if you have a terms aggregation followed by a date histogram aggregation.

m2b commented:
Without looking at other requests, I am hoping this is not too complicated.  You already compute metrics quite nicely from Elastic.  All we need is the capability to define alerts on the panels to work.  Checking against limits, etc.

beardface commented:
@torkelo I had pulled in the es client lib b/c it simplified unmarshalling; but I really dislike having it in here... I'm going to pull it out this morning. Since its pretty bloated for what we need here.

beardface committed:
remove need for olivere/elastic library

beardface commented:
@torkelo Can I get a unit test case that covers this `Also the response handling looks a bit simplistic, does it support multiple levels of bucket aggregations? For example if you have a terms aggregation followed by a date histogram aggregation.` ?

If so, I'm happy to extend it.

Also; I put in a version check for elastic `2.x`... not sure if we need that though; this might support ex `5.x`, but I don't have an active instance to test on... thoughts there?

beardface commented on pkg/tsdb/elasticsearch/elasticsearch.go:
```diff
@@ -0,0 +1,133 @@
+			return result.WithError(err)
+		}
+
+		if esVersion != 2 {
```
Thoughts here?

torkelo commented:
Here is a test case with one terms and one date histogram aggregation, (you can have multiple terms & filter aggregations), each one will nest the result even more, so the response handling needs to handle the recursion. 

````
targets = [{
        refId: 'A',
        metrics: [{type: 'count', id: '1'}, {type: 'avg', field: '@value', id: '4'}],
        bucketAggs: [{type: 'terms', field: 'host', id: '2'}, {type: 'date_histogram', field: '@timestamp', id: '3'}],
      }];
      response =  {
        responses: [{
          aggregations: {
            "2": {
              buckets: [
              {
                "3": {
                  buckets: [
                  { "4": {value: 10}, doc_count: 1, key: 1000},
                  { "4": {value: 12}, doc_count: 3, key: 2000}
                  ]
                },
                doc_count: 4,
                key: 'server1',
              },
              {
                "3": {
                  buckets: [
                  { "4": {value: 20}, doc_count: 1, key: 1000},
                  { "4": {value: 32}, doc_count: 3, key: 2000}
                  ]
                },
                doc_count: 10,
                key: 'server2',
              },
              ]
            }
          }
        }]
      };
```

beardface committed:
support recursive buckets

beardface commented:
@torkelo I added in the test case you provided and extended this to support nested buckets via a recursive processing loop.

I also simplified the types even more and got rid of the need to import anything at all from elastic (or copy paste... which was arguably worse. :) )

beardface committed:
fix nested keys

torkelo commented:
This this PR use the same series name logic (and alias patterns) as the frontend code?

beardface commented:
@torkelo ah; no - missed that.  Adding now.

beardface committed:
add logic for getting preferred names

beardface commented:
@torkelo Ok, I added in logic to compute the preferred names based on the logic from the frontend; did a few tests on my end and is looking good here.

Let me know what you think.

Thanks!

beardface committed:
Get Test Coverage above 80%

beardface committed:
fix test for ci build

torkelo commented:
this is looking really good, will try to merge soon.

torkelo commented on pkg/tsdb/query_context.go:
```diff
@@ -6,9 +6,9 @@ type QueryContext struct {
-	BatchWaits  sync.WaitGroup
+	ResultsChan chan *BatchResult `json:"-"`
+	Lock        sync.RWMutex      `json:"-"`
+	BatchWaits  sync.WaitGroup    `json:"-"`
```
is there a reason for adding these, when is QueryContext ever marshalled to json ?

beardface replied:
Ah, I was initially marshaling the hole thing; but I stopped doing that a while ago; good catch.

beardface committed:
clean up unused code

proffalken commented:
Following, would love to see this merged and working :)

beardface commented:
@torkelo I'm doing some more testing today and it looks like it has problems when you use template variables; but I don't see the template variables getting passed into the context so that I can use it to do replacements...

Is that a known bug?

Is there some way to get the templated variables in `Execute` so that I can do query replacements when sending the request to elasticsearch?

beardface commented:
Upon further thought; that's probably not possible to do easily - or could lead to more confusion when people change their template variable.

Just trying to see what the limitations are and if I missed anything.

beardface committed:
Handling for special characters in queries (quotes)

torkelo commented:
template variables does not need to be supported beside the two built in `$__intervalMS` and `__interval`  (see InfluxDB backend code for reference)

beardface committed:
Replace Interval Variables

beardface commented:
@torkelo based off the code; I presume you meant `$__interval_ms` instead of `$__intervalMS`.

I've updated the logic to handle that case.

torkelo commented:
Yes, I meant that,  great that you fixed that, will merge early next week

beardface committed:
Handle Interval Date Format similar to the JS variant

beardface committed:
Fix interval check and improve query to match frontend

torkelo commented:
Got around to testing this PR today. 

1) Got error that Elasticsearch 5 is not supported

Switched 2 Elasticsearch 2.4 to test further and got this error: 

```
tsdb.HandleRequest() error Failed to get metrics: "{\"error\":{\"root_cause\":[{\"type\":\"index_not_found_exception\",\"reason\":\"no such index\",\"resource.type\":\"index_or_alias\",\"resource.id\":\"metrics-2017.02.22\",\"index\":\"metrics-2017.02.22\"}],\"type\":\"index_not_found_exception\",\"reason\":\"no such index\",\"resource.type\":\"index_or_alias\",\"resource.id\":\"metrics-2017.02.22\",\"index\":\"metrics-2017.02.22\"},\"status\":404}" (404 Not Found)
```

The frontend is building this query
```
{"search_type":"count","ignore_unavailable":true,"index":["metrics-2017.02.21"]}
{"size":0,"query":{"bool":{"filter":[{"range":{"@timestamp":{"gte":"1487670622085","lte":"1487671522085","format":"epoch_millis"}}},{"query_string":{"analyze_wildcard":true,"query":"@metric:logins.count"}}]}},"aggs":{"2":{"date_histogram":{"interval":"10s","field":"@timestamp","min_doc_count":0,"extended_bounds":{"min":"1487670622085","max":"1487671522085"},"format":"epoch_millis"},"aggs":{"1":{"avg":{"field":"@value"}}}}}}
```

Maybe the index patterns in the backend is not using UTC when generating the daily index name? Or some other reason why the backend day is ahead by 1 ?

torkelo commented:
in order for this PR not to get too big and for others to test it more we decided to merge it into a branch, https://github.com/grafana/grafana/tree/alerting-elasticsearch  

please issue bug fixes and further work to that branch.

m2b commented:
I am a bit new on open source.  Should I pull this branch down?  How do I
build it?  Are the binaries?
https://github.com/grafana/grafana/tree/alerting-elasticsearch


On Tue, Feb 21, 2017 at 2:59 AM Torkel Ödegaard <notifications@github.com>
wrote:

> Closed #7551 <https://github.com/grafana/grafana/pull/7551>.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/grafana/grafana/pull/7551#event-970501858>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAvMcYkInV8FycG8OxHgXFYW8Fs5HLjdks5resOlgaJpZM4MADN9>
> .
>
